lib/api/objectHead.js

```diff
@@ -20,38 +23,60 @@ export default function objectHead(authInfo, request, log, callback) {
     log.debug('processing request', { method: 'objectHead' });
     const bucketName = request.bucketName;
     const objectKey = request.objectKey;
-    const metadataValParams = {
+
+    const decryptVidResult = decryptVersionId(request.query);
+    if (decryptVidResult instanceof Error) {
+        log.trace('invalid versionId query', {
+            versionId: request.query.versionId,
+            error: decryptVidResult,
+        });
+        return callback(decryptVidResult);
+    }
+    const versionId = decryptVidResult;
+
+    const mdValParams = {
         authInfo,
         bucketName,
         objectKey,
+        versionId,
         requestType: 'objectHead',
-        log,
     };
 
-    return services.metadataValidateAuthorization(metadataValParams,
+    return metadataValidateBucketAndObj(mdValParams, log,
         (err, bucket, objMD) => {
             const corsHeaders = collectCorsHeaders(request.headers.origin,
                 request.method, bucket);
             if (err) {
-                log.debug('error processing request', {
+                log.debug('error validating request', {
                     error: err,
-                    method: 'metadataValidateAuthorization',
+                    method: 'objectHead',
                 });
                 return callback(err, corsHeaders);
             }
             if (!objMD) {
-                return callback(errors.NoSuchKey, corsHeaders);
+                const err = versionId ? errors.NoSuchVersion : errors.NoSuchKey;
+                return callback(err, corsHeaders);
+            }
+            const verCfg = bucket.getVersioningConfiguration();
+            if (objMD.isDeleteMarker) {
+                const responseHeaders = Object.assign({},
+                    { 'x-amz-delete-marker': true }, corsHeaders);
+                if (!versionId) {
```

Comment:
add comment: if general delete request without versionID and the master object is a delete marker, return NoSuchKey